tests/Unit/Parallel/Test_AlgorithmParallel.cpp

```diff
@@ -319,11 +332,14 @@ struct RemoveInt0 {
         [](const gsl::not_null<int*> count_actions_called) {
           ++*count_actions_called;
         });
-    const bool terminate_algorithm =
-        db::get<Tags::CountActionsCalled>(box) >= 15;
+    if (db::get<Tags::CountActionsCalled>(box) < 150) {
+      Parallel::simple_action<SingletonActions::RestartMe<ParallelComponent>>(
```

Comment:
Is it clear that this action is queued and runs _after_ the algorithm terminates from the return value below? Not sure how this interacts with inlining simple action calls.

Also, could you add a comment here to explain a bit what's going on, in particular how the `150` is chosen?

Reply:
I'm pretty sure it's guaranteed from how our simple and iterable actions work, but I confess it's not terribly obvious. My main understanding of the points we can trust are:
- two actions cannot run simultaneously
- Charm will not preempt an entry method to run another entry method.
- Charm will wait until there are no waiting entry methods to trigger QD and we proceed to the next phase (so performing this termination-restart won't accidentally cause the program to stop early)

So I think this technique is certain to work.

But, yes, I admit the choices for these changes are non-obvious, so I'll add comments.

Reply:
Ok, thanks for explaining. What I'm not entirely clear on is, what happens when the simple action call is inlined before it makes its way to Charm. I suppose at least the algorithm restart won't be inlined, so it might be safe, but @nilsdeppe could you confirm?

Reply:
@moxcodes points:
- correct
- my understanding of this statement is "Charm will not pause an existing entry method to run another". This is correct, there is no support for coroutines and business like that
- correct

We currently do not do any inlining of simple actions, only receive data. Beyond that, I don't know what you mean. QD does not trigger until there are no more messages or entry methods being run. That means that, even if the last action running calls an entry method, QD will not happen until that entry method is completed.